CMakeLists.txt

```diff
@@ -40,17 +44,37 @@ set(ASIO_TAG
     "asio-1-31-0"
     CACHE STRING "asio git tag"
 )
-
 option(ASIO_USE_CPM "Download Asio with CPM instead of FetchContent" ON)
 option(
   ASIO_CPM_FETCHCONTENT_COMPAT
   "Should asio be declared with FetchContent functions to be compatible. This doesn't not allow CPM cache to work."
   OFF
 )
 option(ASIO_NO_DEPRECATED "Disables Asio deprecated interfaces and functionality" ON)
+option(ASIO_SEPARATE_COMPILATION "Uses separately compiled source code for Asio's implementation"
+       ON
+)
+cmake_dependent_option(
+  ASIO_DYN_LINK
+  "Uses separately compiled source code for Asio's implementation, with symbols exported for inclusion as part of a shared library"
+  OFF
+  "ASIO_SEPARATE_COMPILATION"
+  OFF
+)
```

Comment:
Is the extra complexity necessary?
Also should we honor `BUILD_SHARED_LIBS` if the project is top project?

Reply:
Why extra complexity what do you mean? 😅
Since you want to maintain compatibility (`STATIC` as default) how do we can set it with the value of `BUILD_SHARED_LIBS`?
`BUILD_SHARED_LIBS` must control also `ASIO_SEPARATE_COMPILATION` :/

Reply:
My idea is that is this `PROJECT_IS_TOP_LEVEL` and `ASIO_SEPARATE_COMPILATION`, then if `BUILD_SHARED_LIBS` is set, `ASIO_DYN_LINK` should be true by default. This is just an idea, and I guess this is what `BUILD_SHARED_LIBS` is for

Reply:
My initial thought was to not change any _CMake_ (default) behavior because if a user sets `ASIO_SEPARATE_COMPILATION` to `ON` and `BUILD_SHARED_LIBS` to `ON`, it should expects _Asio_ to be a `SHARED` library without specifying `ASIO_DYN_LINK`.
In my opinion, it is not a breaking change because we fully follow _CMake_ behavior. Unfortunately, we now force _Asio_ as a `STATIC` library, which, in my opinion, is not perfect.

Reply:
Yes I see your point, I think we want the same thing. Only my extra condition is `PROJECT_IS_TOP_LEVEL`.
What I want: if a user sets `ASIO_SEPARATE_COMPILATION` to `ON` and `BUILD_SHARED_LIBS` to `ON`, the asio library should be `SHARED`. This should only be the case if `PROJECT_IS_TOP_LEVEL`. Otherwise we keep previous behavior (and I think this is generally something that we want, `BUILD_SHARED_LIBS` should only be listened if the top level project.

Reply:
Why setting `ASIO_DYN_LINK` to the value of `BUILD_SHARED_LIBS` only if `PROJECT_IS_TOP_LEVEL`?
Why not always set it to `BUILD_SHARED_LIBS` (if defined). This is different from version 1.2 but way better for CMake behavior/compatibility
I don't get it :/

Reply:
Because then this would be breaking change to my project that are consuming the repo:
Example:
Main project create library A that is either shared or static. I expect my shared dll that depends on static version of asio to be self-contained, and not depending on external dll for example.
Now if I set `BUILD_SHARED_LIBS` to build my dll, and I follow you your idea, I will end up with 2 dll (my library and asio) and because asio is a private dependency this can become problematic if another dll is available.

Reply:
`INTERFACE` as default! Easy solution 😂

Reply:
Sooo what are the default:
Pseudocode for ASIO_DYN_LINK:
```
if (project_is_top_level) {
  Value is BUILD_SHARED_LIBS if defined otherwise OFF
} else {
  OFF
}
```

Reply:
> Sooo what are the default: Pseudocode for ASIO_DYN_LINK:
> 
> ```
> if (project_is_top_level) {
>   Value is BUILD_SHARED_LIBS if defined otherwise OFF
> } else {
>   OFF
> }
> ```

@OlivierLDff ? 🙌

Reply:
I'm sorry I haven't seen your comment before, and now I see it, absolutely this is the way!

Reply:
@OlivierLDff Done